test/integration/ads_integration_test.cc

```diff
@@ -954,5 +954,24 @@ TEST_P(AdsIntegrationTest, XdsBatching) {
   initialize();
 }
 
+// Validates that listeners can be removed before server start.
+TEST_P(AdsIntegrationTest, ListenerDrainBeforeServerStart) {
+  initialize();
+
+  sendDiscoveryResponse<envoy::api::v2::Cluster>(Config::TypeUrl::get().Cluster,
+                                                 {buildCluster("cluster_0")},
+                                                 {buildCluster("cluster_0")}, {}, "1");
+  sendDiscoveryResponse<envoy::api::v2::ClusterLoadAssignment>(
+      Config::TypeUrl::get().ClusterLoadAssignment, {buildClusterLoadAssignment("cluster_0")},
+      {buildClusterLoadAssignment("cluster_0")}, {}, "1");
+
+  sendDiscoveryResponse<envoy::api::v2::Listener>(
+      Config::TypeUrl::get().Listener, {buildListener("listener_0", "route_config_0")},
+      {buildListener("listener_0", "route_config_0")}, {}, "1");
+
+  // Remove listener.
+  sendDiscoveryResponse<envoy::api::v2::Listener>(Config::TypeUrl::get().Listener, {}, {}, {}, "1");
```

Comment:
Not sure what is the threading model here. Is the `sendDiscoveryResponse` blocked until the removal is done? If not the  test could warn nothing since the listener is not yet starting to remove.

Reply:
Just added a gauge check to verify that the listener was added to active listeners, and removed successfully. The LDS DiscoveryResponse here sends an empty state of the world, which triggers removal.